src/main/java/org/apache/datasketches/filters/bloomfilter/BloomFilter.java

```diff
@@ -0,0 +1,747 @@
+
+    final long seed = mem.getLong(offsetBytes);
+    offsetBytes += Long.BYTES;
+
+    final boolean isEmpty = (flags & EMPTY_FLAG_MASK) != 0;
+
+    final BitArray bitArray = BitArray.heapify(mem.region(offsetBytes, mem.getCapacity() - offsetBytes), isEmpty);
+
+    return new BloomFilter(numHashes, seed, bitArray);
+  }
+
+  /**
+   * Checks if the BloomFilter has processed any items
+   * @return True if the BloomFilter is empty, otherwise False
+   */
+  public boolean isEmpty() { return bitArray_.isEmpty(); }
+  
+  /**
+   * Returns the number of bits in the BloomFilter that are set to 1.
+   * @return The number of bits in use in this filter
+   */
+  public long getBitsUsed() { return bitArray_.getNumBitsSet(); }
+
+  /**
+   * Returns the total number of bits in the BloomFilter.
+   * @return The total size of the BloomFilter
+   */
+  public long getCapacity() { return bitArray_.getCapacity(); }
+
+  /**
+   * Returns the configured number of hash functions for this BloomFilter
+   * @return The number of hash functions to apply to inputs
+   */
+  public short getNumHashes() { return numHashes_; }
+
+  /**
+   * Returns the hash seed for this BloomFilter.
+   * @return The hash seed for this filter
+   */
+  public long getSeed() { return seed_; }
+
+  /**
+   * Returns the percentage of all bits in the BloomFilter set to 1.
+   * @return the percentage of bits in the filter set to 1
+   */
+  public double getFillPercentage() {
+    return (double) bitArray_.getNumBitsSet() / bitArray_.getCapacity();
+  }
+
+  // UPDATE METHODS
+  /**
+   * Updates the filter with the provided long value.
+   * @param item an item with which to update the filter
+   */
+  public void update(final long item) {
+    final long h0 = XxHash.hashLong(item, seed_);
+    final long h1 = XxHash.hashLong(item, h0);
+    updateInternal(h0, h1);
+  }
+
+  /**
+   * Updates the filter with the provided double value. The value is
+   * canonicalized (NaN and infinities) prior to updating.
+   * @param item an item with which to update the filter
+   */
+  public void update(final double item) {
+    // canonicalize all NaN & +/- infinity forms    
+    final long[] data = { Double.doubleToLongBits(item) };
+    final long h0 = XxHash.hashLongArr(data, 0, 1, seed_);
+    final long h1 = XxHash.hashLongArr(data, 0, 1, h0);
+    updateInternal(h0, h1);
+  }
+
+  /**
+   * Updates the filter with the provided String.
+   * The string is converted to a byte array using UTF8 encoding.
+   *
+   * <p>Note: this will not produce the same output hash values as the {@link #update(char[])}
+   * method and will generally be a little slower depending on the complexity of the UTF8 encoding.
+   * </p>
+   *
+   * @param item an item with which to update the filter
+   */
+  public void update(final String item) {
+    if (item == null || item.isEmpty()) { return; }
+    final byte[] strBytes = item.getBytes(StandardCharsets.UTF_8);
```

Comment:
why not do this on toCharArray() skipping conversion?

Reply:
What's going to give the most consistent cross-language behavior?

Reply:
right, for cross-language compatibility this should be better